tower-http/src/timeout/mod.rs

```diff
@@ -0,0 +1,121 @@
+//! Middleware that adds timeouts to request or response bodies.
+//!
+//! Note these middleware differ from [`tower::timeout::Timeout`] which only
+//! adds a timeout to the response future and doesn't consider request bodies.
+
+use tower::BoxError;
+use http_body::Body;
+use pin_project::pin_project;
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+    time::Duration,
+};
+use tokio::time::Sleep;
+
+pub mod request_body;
+pub mod response_body;
+
+#[doc(inline)]
+pub use self::{
+    request_body::{RequestBodyTimeout, RequestBodyTimeoutLayer},
+    response_body::{ResponseBodyTimeout, ResponseBodyTimeoutLayer},
+};
+
+/// An HTTP body with a timeout applied.
+#[pin_project]
+#[derive(Debug)]
+pub struct TimeoutBody<B> {
+    #[pin]
+    inner: B,
+    #[pin]
+    state: State,
+}
+
+impl<B> TimeoutBody<B> {
+    pub(crate) fn new(inner: B, timeout: Duration) -> Self {
+        Self {
+            inner,
+            state: State::NotPolled(timeout),
+        }
+    }
+}
+
+// Only start the timeout after first poll of the body. This enum manages that.
+#[allow(clippy::large_enum_variant)]
+#[pin_project(project = StateProj)]
+#[derive(Debug)]
+enum State {
+    NotPolled(Duration),
+    SleepPending(#[pin] Sleep),
+}
```

Comment:
Not sure how useful this would be. I'm just wondering if users might receive a request and not buffer the body immediately and hit the timeout because they were slow to call the first poll.

Reply:
would it make sense to have a Lazy and non lazy version?

Reply:
It's common to first hit the DB for an auth check with a value from an authorization header before starting to look at the http body. This seems like a good idea to me.

(based on my understanding that the body would usually be polled immediately, except for use cases where the body is streamed)